Add BlockExternalized and PreImageReceived handler events to Agora

TrustHenry opened the pull request:
Add event handlers to be used when the block is externalized and when receiving the preimage.

Call `pushBlock` when the block is externalized.
When receiving a preImage, call `pushPreimage`.

Convert data into JSON format is sent to POST using RestInterfaceClient.

TestNetworkManager does not support it.

Related to #1033

TrustHenry committed:
Add BlockExternalized and PreImageReceived handler events to Agora

Add event handlers to be used when the block is externalized and when receiving the preimage.
This is an implementation of the event handler and has not yet been starting with block reception.
Also, with the preimage reception,

Convert data into JSON format is sent to POST using RestInterfaceClient.

TestNetworkManager does not support it.

AndrejMitrovic commented on source/agora/api/handler/PreImageReceivedHandler.d:
```diff
@@ -0,0 +1,36 @@
+
+    @method(HTTPMethod.POST)
+    @path("/preimage_received")
+    public void pushPreImage (string[string] preimage);
```
Just make the type `Preimage` and let the json serializer do its work. There is no need to do it manually.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,71 @@ public class FullNode : API
+        foreach (BlockExternalizedHandler handler; block_handlers)
+        {
+            runTask({
+                auto jsonStr = block.serializeToJsonString();
```
This is doing manual work which the vibe.d serializer already does for you under the hood.

And secondly, using a map type was unnecessary.

AndrejMitrovic commented:
~Hold on, I will find exactly how to make it use Json again.~

Edit: Yeah, so Json serialization is already in place. Just use the types like `Preimage`, `Block` in the `BlockExternalizedHandler` and `PreImageReceivedHandler`.

And don't use hashmaps, it's wasteful.

TrustHenry commented:
History Agora send 
https://github.com/Geod24/agora/blob/625476b2ba10965ec50fa0c90715b9e930a53b88/source/agora/cli/history/main.d#L94-L96

The code is the code we tested before.

Stoa /push block
https://github.com/bpfkorea/stoa/blob/847c60f82524dc1d91ab302f8037934f0ba1be6c/src/Stoa.ts#L130-L134

If we insert JSON data directly into the `Request.body` , there is no 'block' key, so we need to change our Stoa source.
I wanted to put the block key in the JSON data conversion and I used the hash map as a method.
I removed it because of the double processing part in the @AndrejMitrovic comment.
I need to check if the request.body.data contains the data.
And I applied the suggestion now.

@AndrejMitrovic Thanks for the review comment

AndrejMitrovic commented on source/agora/network/NetworkManager.d:
```diff
@@ -824,6 +825,62 @@ public class NetworkManager
+
+    ***************************************************************************/
+
+    public BlockExternalizedHandler getBlockExternalizedHandler
+        (Address address)
```
```suggestion
    public BlockExternalizedHandler getBlockExternalizedHandler (Address address)
```

TrustHenry replied:
Unfortunately, it's over 80 lines. Are you okay?

AndrejMitrovic replied:
Ah sorry, I didn't realize. No it's fine then.

But we should probably think about reducing the length of our symbol names.

TrustHenry replied:
I wanted to reduce it, but I can't think of anything. Could you recommend it?
https://github.com/bpfkorea/agora/issues/1033
Actually, I decided to follow the example of the issue.

AndrejMitrovic replied:
It's fine for now. Let's get this through and think about refactoring later.

TrustHenry replied:
That's a good idea.

AndrejMitrovic commented on source/agora/network/NetworkManager.d:
```diff
@@ -824,6 +825,62 @@ public class NetworkManager
+
+    ***************************************************************************/
+
+    public PreImageReceivedHandler getPreimageReceivedHandler
+        (Address address)
```
```suggestion
    public PreImageReceivedHandler getPreimageReceivedHandler (Address address)
```

AndrejMitrovic replied:
Ignore this suggestion.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -149,6 +157,18 @@ public class FullNode : API
+        // Make `BlockExternalizedHandler` from config
+        foreach (block_externalize_address;
+            config.event_handlers.block_externalized_handler_addresses)
+            this.block_handlers ~= this.network.getBlockExternalizedHandler
+                (block_externalize_address);
```
```suggestion
            this.block_handlers ~= this.network
                .getBlockExternalizedHandler(block_externalize_address);
```

There is a reason I suggest this. It's to make it easier to find function calls across the codebase.

TrustHenry replied:
Yes, it looks much better.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -149,6 +157,18 @@ public class FullNode : API
+        // Make `PreImageReceivedHandler` from config
+        foreach (preimage_updated_address;
+            config.event_handlers.preimage_updated_handler_addresses)
+            this.preimage_handlers ~= this.network.getPreimageReceivedHandler
+                (preimage_updated_address);
```
```suggestion
            this.preimage_handlers ~= this.network
                .getPreimageReceivedHandler(preimage_updated_address);
```

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,69 @@ public class FullNode : API
+
+    public void pushBlock (const ref Block block) @safe
+    {
+        foreach (BlockExternalizedHandler handler; block_handlers)
```
```suggestion
        foreach (BlockExternalizedHandler handler; this.block_handlers)
```

TrustHenry replied:
Good catch!

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,69 @@ public class FullNode : API
+                catch (Exception e)
+                {
+                    log.error(
+                        "Error sending block height #", block.header.height
```
Will this actually work? I didn't realize it was supported. I thought we had to use `log.error("block height #{}", height)` - the formatting string.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,69 @@ public class FullNode : API
+                {
+                    log.error(
+                        "Error sending block height #", block.header.height
+                        ," to \"", this.config.node.address, "\": ", e);
```
This is wrong. It's not sending the block to itself.

AndrejMitrovic replied:
Also in the future you can use this:

```
` to"`
```

This is simpler than using `\"`.

TrustHenry replied:
It's fixed now.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,69 @@ public class FullNode : API
+
+    public void pushPreImage (const ref PreImageInfo preImage) @safe
+    {
+        foreach (PreImageReceivedHandler handler; preimage_handlers)
```
```suggestion
        foreach (PreImageReceivedHandler handler; this.preimage_handlers)
```

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,69 @@ public class FullNode : API
+                    log.error(
+                        "Error sending preImage #",
+                        " (enroll_key: ", preImage.enroll_key, ") to \"",
+                        this.config.node.address, "\": ", e);
```
Again wrong destination address.

TrustHenry replied:
Yes, that's right.
This is the remnant used before adding config.

AndrejMitrovic commented:
> If we insert JSON data directly into the Request.body , there is no 'block' key, so we need to change our Stoa source.

So just add another parameter to `pushBlock` then.

```D
public void pushBlock (Hash hash, const Block block);
```

TrustHenry commented:
Associative Arrays was used to print the target address of the event handler in the event of an error.

TrustHenry commented:
> > If we insert JSON data directly into the Request.body , there is no 'block' key, so we need to change our Stoa source.
> 
> So just add another parameter to `pushBlock` then.
> 
> ```d
> public void pushBlock (Hash hash, const Block block);
> ```

Post in 'pushBlock' We will just push JSON block data from body and fix respond from stoa.
https://github.com/bpfkorea/stoa/pull/40

MichaelKim20 commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,66 @@ public class FullNode : API
+
+    public void pushBlock (const ref Block block) @safe
+    {
+        foreach (handler_address; this.block_handlers.byKey)
```
I think this would be better.

```suggestion
        foreach (address, handler; this.block_handlers)
```
Link to reference)
https://dlang.org/spec/hash-map.html#aa_example_iteration

TrustHenry replied:
Good! This looks better.

MichaelKim20 commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,66 @@ public class FullNode : API
+
+    public void pushPreImage (const ref PreImageInfo preImage) @safe
+    {
+        foreach (handler_address; this.preimage_handlers.byKey)
```
I think this would be better.

```suggestion
        foreach (address, handler; this.preimage_handlers)
```
Link to reference)
https://dlang.org/spec/hash-map.html#aa_example_iteration

MichaelKim20 commented on source/agora/api/handler/BlockExternalizedHandler.d:
```diff
@@ -0,0 +1,38 @@
+
+public interface BlockExternalizedHandler
+{
+// The REST generator requires @safe methods
+@safe:
```
Is this really necessary?

MichaelKim20 commented on source/agora/api/handler/PreImageReceivedHandler.d:
```diff
@@ -0,0 +1,38 @@
+
+public interface PreImageReceivedHandler
+{
+// The REST generator requires @safe methods
+@safe:
```
Is this really necessary?

AndrejMitrovic replied:
> Is this really necessary?

Don't you see the comment? It says `REST generator requires @safe`

MichaelKim20 reviewed (commented):
I think I need to check a few things.
The rest is good.

MichaelKim20 reviewed (approved):
LGTM

linked0 commented on source/agora/network/NetworkManager.d:
```diff
@@ -824,6 +826,62 @@ public class NetworkManager
+        settings.httpClientSettings.connectTimeout = this.node_config.timeout;
+        settings.httpClientSettings.readTimeout = this.node_config.timeout;
+
+        return new RestInterfaceClient!PreImageReceivedHandler(address);
```
Don't you need to set `settings` instead of the `address`?

TrustHenry replied:
Not a single value.
Multiple `PreImageRecoveredHandler settings create a new instance `RestInterfaceClient`.

AndrejMitrovic replied:
He means you should write this:
```suggestion
        return new RestInterfaceClient!PreImageReceivedHandler(settings);
```

TrustHenry replied:
Oh, I didn't see it.
That's right. It's wrong.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -106,6 +108,12 @@ public class FullNode : API
     private immutable Block genesis_block;
 
+    /// Block Externalized Handler list
+    protected BlockExternalizedHandler[string] block_handlers;
```
```suggestion
    protected BlockExternalizedHandler[Address] block_handlers;
```

TrustHenry replied:
Good! Use the defined type.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -106,6 +108,12 @@ public class FullNode : API
+    protected BlockExternalizedHandler[string] block_handlers;
+
+    /// PreImage Received Handler list
+    protected PreImageReceivedHandler[string] preimage_handlers;
```
```suggestion
    protected PreImageReceivedHandler[Address] preimage_handlers;
```

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,65 @@ public class FullNode : API
+
+    ***************************************************************************/
+
+    public void pushBlock (const ref Block block) @safe
```
Why is this public?

And why is it never called?

TrustHenry replied:
I think it will be called outside the FullNode class.

TrustHenry replied:
I'm preparing for the next PR.

AndrejMitrovic replied:
> I think it will be called outside the FullNode class.

???

TrustHenry replied:
I'm thinking about when to call.
BlockStorage.saveblock thinks it's the starting point for `pushBlock` delegate events.

AndrejMitrovic replied:
You want BlockStorage to be the trigger point that calls FullNode.pushBlock? You want to add FullNode as an instance to the BlockStorage class?

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +486,65 @@ public class FullNode : API
+
+    ***************************************************************************/
+
+    public void pushPreImage (const ref PreImageInfo preImage) @safe
```
```suggestion
    public void pushPreImage (const ref PreImageInfo pre_image) @safe
```

Styling rules.

AndrejMitrovic replied:
Or just `preimage`

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -149,6 +157,18 @@ public class FullNode : API
         this.exception = new RestException(
             400, Json("The query was incorrect"), string.init, int.init);
+
+        // Make `BlockExternalizedHandler` from config
+        foreach (block_externalize_address;
+            config.event_handlers.block_externalized_handler_addresses)
+            this.block_handlers[block_externalize_address] = this.network
+                .getBlockExternalizedHandler(block_externalize_address);
+
+        // Make `PreImageReceivedHandler` from config
+        foreach (preimage_updated_address;
+            config.event_handlers.preimage_updated_handler_addresses)
+            this.preimage_handlers[preimage_updated_address] = this.network
+                .getPreimageReceivedHandler(preimage_updated_address);
```
This can be simplified because the variables are not re-used outside the foreach statement.

```suggestion
        // Make `BlockExternalizedHandler` from config
        foreach (address;
            config.event_handlers.block_externalized_handler_addresses)
            this.block_handlers[address] = this.network.getBlockExternalizedHandler(address);

        // Make `PreImageReceivedHandler` from config
        foreach (address;
            config.event_handlers.preimage_updated_handler_addresses)
            this.preimage_handlers[address] = this.network.getPreimageReceivedHandler(address);
```

TrustHenry replied:
Good! The code is more concise.

AndrejMitrovic commented:
@TrustHenry you can do something like this:

```diff
diff --git a/source/agora/node/FullNode.d b/source/agora/node/FullNode.d
index 035a2205..a84249c9 100644
--- a/source/agora/node/FullNode.d
+++ b/source/agora/node/FullNode.d
@@ -154,7 +154,8 @@ public class FullNode : API
         this.enroll_man = this.getEnrollmentManager(config.node.data_dir,
             config.node, params);
         this.ledger = new Ledger(config.node, params, this.utxo_set,
-            this.storage, this.enroll_man, this.pool, onValidatorsChanged);
+            this.storage, this.enroll_man, this.pool, &this.pushBlock,
+            onValidatorsChanged);
         this.exception = new RestException(
             400, Json("The query was incorrect"), string.init, int.init);
 
@@ -499,7 +500,7 @@ public class FullNode : API
 
     ***************************************************************************/
 
-    public void pushBlock (const ref Block block) @safe
+    private void pushBlock (const ref Block block) @safe
     {
         foreach (address, handler; this.block_handlers)
         {
diff --git a/source/agora/node/Ledger.d b/source/agora/node/Ledger.d
index edc1e786..de2e7c27 100644
--- a/source/agora/node/Ledger.d
+++ b/source/agora/node/Ledger.d
@@ -79,6 +79,7 @@ public class Ledger
     /// If not null call this delegate if the validator set changed after
     /// a block was externalized
     private void delegate () nothrow @safe onValidatorsChanged;
+    private void delegate (const ref Block) nothrow @safe onAcceptedBlock;
 
     /// Parameters for consensus-critical constants
     private immutable(ConsensusParams) params;
@@ -103,6 +104,7 @@ public class Ledger
         NodeConfig node_config, immutable(ConsensusParams) params,
         UTXOSet utxo_set, IBlockStorage storage,
         EnrollmentManager enroll_man, TransactionPool pool,
+        void delegate (const ref Block) nothrow @safe onAcceptedBlock,
         void delegate () nothrow @safe onValidatorsChanged = null)
     {
         this.node_config = node_config;
@@ -111,6 +113,7 @@ public class Ledger
         this.storage = storage;
         this.enroll_man = enroll_man;
         this.pool = pool;
+        this.onAcceptedBlock = onAcceptedBlock;
         this.onValidatorsChanged = onValidatorsChanged;
         if (!this.storage.load(GenesisBlock))
             assert(0);
@@ -275,6 +278,8 @@ public class Ledger
 
         if (this.onValidatorsChanged !is null && validators_changed)
             this.onValidatorsChanged();
+
+        this.onAcceptedBlock(block);
     }
 
     /***************************************************************************
```

And for preimages you can call `pushPreImage` in the `receivePreimage` method.

TrustHenry committed:
push to config setting when receiving preImage

config `preimage_updated_handler_addresses`

TrustHenry committed:
Call the pushBlock when block externalized

Call the delegate at the time of Leder.addValidatedBlock.
PushBlock once if there is block height 0 when Fullnode is starting.

TrustHenry commented:
@AndrejMitrovic Thank you. I applied the suggestion now.
 and added two new commits.

Call pushBlock when the block is externalized.
When receiving a preImage, call pushPreimage.

TrustHenry commented:
Ready to review!

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -80,6 +80,10 @@ public class Ledger
     /// a block was externalized
     private void delegate () nothrow @safe onValidatorsChanged;
 
+    /// If not null call this delegate
+    /// A block was externalized
```
```suggestion
    /// If not null call this delegate when a new block was added to the ledger
```

It does not necessarily mean "externalized", because this delegate will be called when a node is in catch-up mode. For example when calling `getBlocksFrom`.

This is why I don't like to call the handler "BlockExternalizedHandler".

However we don't have to worry about naming now, we can change it later.

TrustHenry replied:
I also considered not focusing on Block Externalized. I think a few word have been added.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -94,6 +98,8 @@ public class Ledger
             enroll_man = the enrollmentManager
             pool = the transaction pool
+            onAcceptedBlock = optional delegate to call
+                              when a block was externalized
```
```suggestion
                              when a block was added to the ledger
```

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -103,6 +109,7 @@ public class Ledger
         NodeConfig node_config, immutable(ConsensusParams) params,
         UTXOSet utxo_set, IBlockStorage storage,
         EnrollmentManager enroll_man, TransactionPool pool,
+        void delegate (const ref Block) @safe onAcceptedBlock = null,
```
This should not be an optional parameter (remove `null`).

AndrejMitrovic replied:
The FullNode always provides a delegate, and a Validator inherits the full node so it has the same behavior. So it is not an optional parameter.

TrustHenry replied:
That's right. I was concerned about that.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -275,6 +284,9 @@ public class Ledger
         if (this.onValidatorsChanged !is null && validators_changed)
             this.onValidatorsChanged();
+
+        if (this.onAcceptedBlock !is null)
```
It should never be null.

AndrejMitrovic reviewed (changes requested):
Requested a few small changes.

This is a good first step, however it would be best to integrate this code into the NetworkManager in the future. Because it has request retrying support, and the current code does not. However, care needs to be taken (we should never ban event handlers, the ban manager should not be used).

TrustHenry commented:
Thank you. I applied now all of @AndrejMitrovic's suggestions.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -94,6 +97,8 @@ public class Ledger
             storage = the block storage
             enroll_man = the enrollmentManager
             pool = the transaction pool
+            onAcceptedBlock = optional delegate to call
```
```suggestion
            onAcceptedBlock = delegate to call
```

AndrejMitrovic replied:
Ah @TrustHenry I didn't realize this was also used in unittests.

Maybe making it optional was actually a good idea. I'm sorry to give you more work, it might be worth reverting the change and making it optional (with the `!is null` check).

TrustHenry replied:
The code has been reverted. OnAcceptedBlock is set to Optional. with the !is null check
Thanks for your review.

AndrejMitrovic commented on source/agora/node/Ledger.d:
```diff
@@ -538,7 +548,7 @@ version (unittest)
                 new MemBlockStorage(blocks),
                 new EnrollmentManager(":memory:", config.key_pair, params),
-                new TransactionPool(":memory:"));
+                new TransactionPool(":memory:"), (const ref Block) @safe {});
```
This is the part I didn't see before. Thanks for making it clearer to me @TrustHenry.

AndrejMitrovic reviewed (approved):
Approved, there were just a few issues.

TrustHenry commented:
It's kind of weird.
I didn't see all the CI tests in this PR turn green.
One of the CI tests continues to lag and fails to output logs.

TrustHenry commented:
@Geod24 ping

MichaelKim20 commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +495,65 @@ public class FullNode : API
+            runTask({
+                try
+                {
+                    handler.pushBlock(block);
```
I have a question.
Will this value(block) be valid here?

https://github.com/bpfkorea/agora/blob/8056af52285f23eb6937102c3bd7662c013eec2c/source/agora/consensus/protocol/Nominator.d#L395

TrustHenry replied:
I'm not sure, 
I think he's asking if the block is consensused block. Is this what you meant?
And do you want to know if here the `pushBlock` call point is okay?

AndrejMitrovic replied:
`pushBlock` is called when a **validated** block is added to the ledger. It is unrelated to nomination.

MichaelKim20 replied:
@AndrejMitrovic 
When a local variable created by another function is approached within the runTask,
The local variable's contents are missing.
How should I solve this?
I upload the sample code I made together.
Please.
https://gist.github.com/MichaelKim20/0c626a54dfd1f68593fb886da059c20b

AndrejMitrovic replied:
Ah that's right, we should not use `ref` here. The reference may be out of scope when the task starts.

@TrustHenry you will need to remove `ref` from the function parameter.

TrustHenry replied:
Yes, it certainly does.
Thank you.
I will apply `pushPreImage` as well.

TrustHenry commented:
CI test was performed after the rebase(https://github.com/bpfkorea/agora/pull/1046) but failed to pass ubuntu-18.04 ldc-master.
Previously, errors only occurred in macOS-10.15,ldc-1.22.0 

https://github.com/bpfkorea/agora/pull/1038/checks?check_run_id=896701409

````
+ dchatty=1
+ timeout -s SEGV 8m ./build/agora-unittests
./build/agora-unittests(+0x772b47)[0x5608d25c2b47]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x128a0)[0x7f6f86a6e8a0]
/lib/x86_64-linux-gnu/libpthread.so.0(pthread_cond_wait+0x243)[0x7f6f86a699f3]
./build/agora-unittests(+0x772f20)[0x5608d25c2f20]
./build/agora-unittests(+0x7602ba)[0x5608d25b02ba]
./build/agora-unittests(+0x760028)[0x5608d25b0028]
./build/agora-unittests(+0x1b2ee3)[0x5608d2002ee3]
./build/agora-unittests(+0x1b1960)[0x5608d2001960]
./build/agora-unittests(+0x7729d1)[0x5608d25c29d1]
./build/agora-unittests(+0x77e4cb)[0x5608d25ce4cb]
./build/agora-unittests(+0x77e3f3)[0x5608d25ce3f3]
./build/agora-unittests(+0x77e24e)[0x5608d25ce24e]
./build/agora-unittests(+0x1e6865)[0x5608d2036865]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f6f860d6b97]
./build/agora-unittests(+0x7c65a)[0x5608d1ecc65a]
##[error]The operation was canceled.
```

TrustHenry commented:
The run time of the build&test run task stopped at 29m 21s.

TrustHenry commented:
@Geod24 Ready to review.

TrustHenry commented on source/agora/node/FullNode.d:
```diff
@@ -466,4 +495,65 @@ public class FullNode : API
+            runTask({
+                try
+                {
+                    handler.pushPreImage(pre_image);
```
There is a problem that I think.
When matching `PreImageInfo` information with `enrolled enrolment` in `Stoa`
If there are multiple `enrolled enrolments` with the same UTXO_key,
We have to think about which `PreImageInfo` matches which of the two other `Enrollments`.
Because there is no information about `PreImageCycle` in `Stoa`.
Therefore, we need to make sure that `Hashing` matches `Random seed` of `Enrollment` by doing `Distance` of `PreImageInfo`.
However, it can be inefficient if Stoa does Hashing every time it receives `Preimage`.

AndrejMitrovic replied:
> If there are multiple enrolled enrolments with the same UTXO_key,

Multiple enrollments with the same UTXO? That shouldn't be possible.

linked0 replied:
> Multiple enrollments with the same UTXO? That shouldn't be possible.

He means continuous enrollments with the same UTXO. He's worried about the situation where some applications could proceed during a period of more than one enrollment. So he wants to know which enrollment is the one for a specific pre-image among continuous enrollments.

TrustHenry replied:
> Multiple enrollments with the same UTXO? That shouldn't be possible.

@AndrejMitrovic  Hold on plz ,,, Most likely to use the same UTXO.

@linked0  That's right.

TrustHenry replied:
Here's what @linked0  describes as a `continuous enrollments`
For example, multiple enrollments(continuous enrollments) mean to the PreImageCycle 0 enrollment and the PreImageCycle 1 enrollment.

TrustHenry replied:
The word `multiple enrollments` seems to cause confusion, so I'll call it `continuous  enrollments`

AndrejMitrovic replied:
But they cannot be active at the same block height, right?

TrustHenry replied:
> But they cannot be active at the same block height, right?

Yes, I'd like to believe that way.

AndrejMitrovic replied:
So is there a problem here or not? Is the PR ready to go?

TrustHenry replied:
> So is there a problem here or not? Is the PR ready to go?

I think Stoa needs to figure it out.

AndrejMitrovic commented:
Can you please rebase? I wonder if the LocalRest fix is related to the timeout problem..

AndrejMitrovic commented:
This is strange, now there is a different error.

AndrejMitrovic commented:
SIGBUS instead of SIGSEGV, which means it was not the `timeout` which killed the binary.

TrustHenry commented:
It's strange, but I'll try again.

TrustHenry commented:
@Geod24 do you want to look into it more, or it looks good to you?
